File: functions/src/main/java/com/tunjid/androidbootstrap/functions/TransformingSequentialList.java
package com.tunjid.androidbootstrap.functions;

import java.util.AbstractSequentialList;
import java.util.Iterator;
import java.util.List;
import java.util.ListIterator;

public final class TransformingSequentialList<F, T> extends AbstractSequentialList<T> {

    private final List<F> fromList;

    private final Function<? super T, ? extends F> toFunction;
    private final Function<? super F, ? extends T> fromFunction;

    public static <F, T> List<T> transform(List<F> fromList,
                                           Function<? super F, ? extends T> fromFunction) {
        return transform(fromList, fromFunction, null);
    }

    public static <F, T> List<T> transform(List<F> fromList,
                                           Function<? super F, ? extends T> fromFunction,
                                           Function<? super T, ? extends F> toFunction) {
        return new TransformingSequentialList<>(fromList, fromFunction, toFunction);
    }

    private TransformingSequentialList(List<F> fromList,
                                       Function<? super F, ? extends T> fromFunction,
                                       Function<? super T, ? extends F> toFunction) {
        this.fromList = fromList;
        this.toFunction = toFunction;
        this.fromFunction = fromFunction;
    }

    /**
     * The default implementation inherited is based on iteration and removal of each element which
     * can be overkill. That's why we forward this call directly to the backing list.
     */
    @Override
    public void clear() { fromList.clear(); }

    @Override
    public int size() { return fromList.size(); }

    @Override
    public ListIterator<T> listIterator(final int index) {
        final ListIterator<F> fromListIterator = fromList.listIterator(index);

        return new TransformedListIterator<F, T>(fromListIterator) {
            @Override
            T transform(F from) { return fromFunction.apply(from); }

            @Override
            public void set(T element) {
                if (toFunction == null) throw new UnsupportedOperationException();
                else fromListIterator.set(toFunction.apply(element));
            }

            @Override
            public void add(T element) {
                if (toFunction == null) throw new UnsupportedOperationException();
                else fromListIterator.add(toFunction.apply(element));
            }
        };
    }

    abstract static class TransformedListIterator<F, T> extends TransformedIterator<F, T>
            implements ListIterator<T> {

        TransformedListIterator(ListIterator<? extends F> backingIterator) {
            super(backingIterator);
        }

        private ListIterator<? extends F> backingIterator() {
            return cast(backingIterator);
        }

        @Override
        public final boolean hasPrevious() { return backingIterator().hasPrevious(); }

        @Override
        public final T previous() { return transform(backingIterator().previous()); }

        @Override
        public final int nextIndex() { return backingIterator().nextIndex(); }

        @Override
        public final int previousIndex() { return backingIterator().previousIndex(); }

        @Override
        public void set(T element) { throw new UnsupportedOperationException(); }

        @Override
        public void add(T element) { throw new UnsupportedOperationException(); }

        private static <T> ListIterator<T> cast(Iterator<T> iterator) {
            return (ListIterator<T>) iterator;
        }
    }

    abstract static class TransformedIterator<F, T> implements Iterator<T> {
        final Iterator<? extends F> backingIterator;

        TransformedIterator(Iterator<? extends F> backingIterator) {
            this.backingIterator = backingIterator;
        }

        abstract T transform(F from);

        @Override
        public final boolean hasNext() {
            return backingIterator.hasNext();
        }

        @Override
        public final T next() {
            return transform(backingIterator.next());
        }

        @Override
        public final void remove() {
            backingIterator.remove();
        }
    }
}

File: functions/src/main/java/com/tunjid/androidx/functions/collections/TransformingSequentialList.kt
package com.tunjid.androidx.functions.collections

import java.util.*

internal class TransformingSequentialList<F, T>(private val fromList: MutableList<F>,
                                                private val fromFunction: (F) -> T,
                                                private val toFunction: ((T) -> F)?) : AbstractSequentialList<T>() {

    override val size: Int
        get() = fromList.size

    /**
     * The default implementation inherited is based on iteration and removal of each element which
     * can be overkill. That's why we forward this call directly to the backing list.
     */
    override fun clear() = fromList.clear()

    override fun listIterator(index: Int): MutableListIterator<T> {
        val fromListIterator = fromList.listIterator(index)

        return object : TransformedListIterator<F, T>(fromListIterator, toFunction) {
            override fun transform(from: F): T = fromFunction(from)

            override fun set(element: T) =
                if (toFunction == null) throw UnsupportedOperationException()
                else fromListIterator.set(toFunction.invoke(element))

            override fun add(element: T) =
                if (toFunction == null) throw UnsupportedOperationException()
                else fromListIterator.add(toFunction.invoke(element))
        }
    }

    internal abstract class TransformedListIterator<F, T>(
        private val backingIterator: MutableListIterator<F>,
        private val toFunction: ((T) -> F)?
    ) : TransformedIterator<F, T>(backingIterator), MutableListIterator<T> {

        override fun hasPrevious(): Boolean = backingIterator.hasPrevious()

        override fun previous(): T = transform(backingIterator.previous())

        override fun nextIndex(): Int = backingIterator.nextIndex()

        override fun previousIndex(): Int = backingIterator.previousIndex()

        override fun set(element: T) =
            if (toFunction != null) backingIterator.set(toFunction.invoke(element))
            else throw UnsupportedOperationException()

        override fun add(element: T) =
            if (toFunction != null) backingIterator.set(toFunction.invoke(element))
            else throw UnsupportedOperationException()
    }

    internal abstract class TransformedIterator<F, T>(private val backingIterator: MutableIterator<F>) : MutableIterator<T> {

        internal abstract fun transform(from: F): T

        override fun hasNext(): Boolean = backingIterator.hasNext()

        override fun next(): T = transform(backingIterator.next())

        override fun remove() = backingIterator.remove()
    }
}
